subsys/net/ip/net_context.c

```diff
@@ -566,7 +566,9 @@ int net_context_bind(struct net_context *context, const struct sockaddr *addr,
 
 		if (IS_ENABLED(CONFIG_NET_OFFLOAD) &&
 		    net_if_is_ip_offloaded(iface)) {
-			net_context_set_iface(context, iface);
+			if (!net_context_is_bound_to_iface(context)) {
+				net_context_set_iface(context, iface);
+			}
```

Comment:
This looks wrong. Lets say that you have native "ethernet" interface and offloaded "wifi" interface. When creating socket, you have bound it (with SO_BINDTODEVICE) to "ethernet", but `iface` could point now to "wifi" because of the `addr` that you pass to `net_context_bind`. So at the end you will leave "ethernet" in `context->iface`, but still call "wifi" drivers with a call to `net_offload_bind` below.

Maybe `context->iface` needs to be used directly when `net_context_is_bound_to_iface(context) == true` instead of figuring out `iface` by from `addr`?

Reply:
Yes, you're right, the current approach will not work with offloaded devices, good catch! I'll rework it according to your proposal, which sounds like a good solution for the issue.